### PR TITLE
Add step 'When I wait for 'responseCode' response code for HTTP httpM…

### LINKED ISSUE #546: No ability to wait for a response code
Sometimes, after uploading an asset (a file), some processing happens on server side, and the resource is not available for downloading (response code is 404).

It would be useful to have something like:
When I wait for <response_code> response code for HTTP <http_method> <resource_url> during <duration_time>

### PR DESCRIPTION
**Pull request by ngrudnitsky**
…ethod request for a resource with the URL 'url' for 'duration' duration retrying retryTimes times' Closes #546

### CODE REVIEW

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 1]
```diff
@@ -149,6 +156,14 @@ void testSetCustomRequestConfigNoField()
         assertEquals("nonExistentField", exception.getMessage());
     }
 
+    @Test
+    void testWaitForResponse() throws IOException
+    {
+        httpRequestSteps.waitForResponse(RESPONSE_CODE, HttpMethod.GET, URL, DURATION, RETRY_TIMES);
+        verify(httpRequestExecutor).executeHttpRequest(eq(HttpMethod.GET), eq(URL), eq(Optional.empty()),
+                any(), any(WaitMode.class));
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:164]: do not use `any` to verify expected successful invocations, use `eq`, `argThat` and etc.

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 2]
```diff
@@ -47,6 +51,9 @@
     private static final String RELATIVE_URL = "/relativeUrl";
     private static final String CONTENT = "content";
     private static final String CR_LF = "\r\n";
+    private static final Duration DURATION = Duration.ofSeconds(10);
+    private static final int RESPONSE_CODE = 200;
+    private static final int RETRY_TIMES = 10;
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:56]: constants used once could be inlined

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 3]
```diff
@@ -202,6 +204,39 @@ public void setCustomRequestConfig(ExamplesTable configItems) throws ReflectiveO
...
+     *                     <li>PATCH</li>
+     *                     </ul>
+     * @param url          for example https://www.vividus.org/sitemap.xml
+     * @param duration     Time duration to wait
+     * @param retryTimes   How many times request will be retried; duration/retryTimes=timeout between requests
+     * @throws IOException If an input or output exception occurred
+     */
+    @When("When I wait for '$responseCode' response code for HTTP $httpMethod request for a resource "
+            + "with the URL '$url' for '$duration' duration retrying $retryTimes times")
+    public void waitForResponse(int responseCode, HttpMethod httpMethod, String url, Duration duration,
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:233]: `waitForResponse` -> `waitForResponseCode`

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 4]
```diff
@@ -202,6 +204,39 @@ public void setCustomRequestConfig(ExamplesTable configItems) throws ReflectiveO
         httpTestContext.putRequestConfig(requestConfigBuilder.build());
     }
 
+    /**
+     * Step sends HTTP request for the given <b>url</b> and waits for a specified amount of time
+     * until HTTP response code is equal to what is expected and stores HTTP response into Scenario-level variable.<br>
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:209]: ~and stores HTTP response into Scenario-level variable~

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 5]
```diff
@@ -202,6 +204,39 @@ public void setCustomRequestConfig(ExamplesTable configItems) throws ReflectiveO
...
+     *                     <li>TRACE</li>
+     *                     <li>PATCH</li>
+     *                     </ul>
+     * @param url          for example https://www.vividus.org/sitemap.xml
+     * @param duration     Time duration to wait
+     * @param retryTimes   How many times request will be retried; duration/retryTimes=timeout between requests
+     * @throws IOException If an input or output exception occurred
+     */
+    @When("When I wait for '$responseCode' response code for HTTP $httpMethod request for a resource "
+            + "with the URL '$url' for '$duration' duration retrying $retryTimes times")
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:232]: @When("I wait for `$responseCode` response code for HTTP $httpMethod request to resource "
           + "`$resourceUrl` for `$duration` duration retrying $retryTimes times")

Comment (valfirst) [reply vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:232]: Also what about making this step a bit more universal:

```
@When("I wait for response code `$responseCode` for `$duration` duration retrying $retryTimes times"
        + "$stepsToExecute")
```
?

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 6]
```diff
@@ -202,6 +207,42 @@ public void setCustomRequestConfig(ExamplesTable configItems) throws ReflectiveO
...
+        );
+    }
+
+    private boolean isResponseCodeIsEqualToExpected(HttpResponse response, int expectedResponseCode)
+    {
+        if (response == null)
+        {
+            return false;
+        }
+        int statusCode = response.getStatusCode();
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:242]: the variable can be inlined

File: vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java [hunk 7]
```diff
@@ -202,6 +207,42 @@ public void setCustomRequestConfig(ExamplesTable configItems) throws ReflectiveO
...
+        return statusCode == expectedResponseCode;
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/HttpRequestSteps.java:243]: in general the full method can be like:
```java
return response != null && response.getStatusCode() == expectedResponseCode;
```

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 8]
```diff
@@ -149,6 +155,18 @@ void testSetCustomRequestConfigNoField()
...
     }
 
+    @Test
+    void testWaitForResponseCode()
+    {
+        int retryTimes = 5;
+        int responseCode = 200;
+        SubSteps stepsToExecute = mock(SubSteps.class);
+        when(httpTestContext.getResponse())
+                .thenReturn(new HttpResponse());
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:165]: this peace of code can be put on the previous line

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 9]
```diff
@@ -149,6 +155,18 @@ void testSetCustomRequestConfigNoField()
...
+        httpRequestSteps.waitForResponseCode(responseCode, Duration.ofSeconds(10), retryTimes, stepsToExecute);
+        verify(stepsToExecute, times(retryTimes)).execute(Optional.empty());
+    }
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:168]: missing tests for cases:
- response == null
- response.code == expected response code

File: vividus-tests/src/main/resources/story/integration/HttpRequest.story [hunk 10]
```diff
@@ -25,3 +25,8 @@ Then a JSON element by the JSON path '$.cookies' is equal to '{"vividus-cookie":
 Scenario: Verify HTTP cookies are cleared
 When I send HTTP GET to the relative URL '/cookies'
 Then a JSON element by the JSON path '$.cookies' is equal to '{}'
+
+Scenario: I wait for response code $responseCode for $duration duration retrying $retryTimes times $stepsToExecute
+When I wait for response code 200 for PT10S duration retrying 10 times
+|step                                                                                           |
+|When I issue a HTTP GET request for a resource with the URL '${vividus-test-site-url}/api/delayed-response'|
```
Comment (Yauhenda) [inline vividus-tests/src/main/resources/story/integration/HttpRequest.story:32]: Please add an assertion `Then the response code is equal to '200'`

File: vividus-tests/src/main/resources/story/integration/HttpRequest.story [hunk 11]
```diff
@@ -25,3 +25,8 @@ Then a JSON element by the JSON path '$.cookies' is equal to '{"vividus-cookie":
 Scenario: Verify HTTP cookies are cleared
 When I send HTTP GET to the relative URL '/cookies'
 Then a JSON element by the JSON path '$.cookies' is equal to '{}'
+
+Scenario: I wait for response code $responseCode for $duration duration retrying $retryTimes times $stepsToExecute
+When I wait for response code 200 for PT10S duration retrying 10 times
+|step                                                                                           |
```
Comment (Yauhenda) [inline vividus-tests/src/main/resources/story/integration/HttpRequest.story:31]: Please align the `|`  sign to the same indent  as next line to make code more beautiful

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 12]
```diff
@@ -149,6 +157,35 @@ void testSetCustomRequestConfigNoField()
...
+
+    @Test
+    void testWaitForResponseCodeWhenResponseIsNull()
+    {
+        SubSteps stepsToExecute = mock(SubSteps.class);
+        when(httpTestContext.getResponse()).thenReturn(null);
+        httpRequestSteps.waitForResponseCode(RESPONSE_CODE, Duration.ofSeconds(10), RETRY_TIMES, stepsToExecute);
+        verify(stepsToExecute, times(RETRY_TIMES)).execute(Optional.empty());
+    }
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:176]: this test and the previous one can be combined and data-driven approach can be applied: https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources-MethodSource

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 13]
```diff
@@ -149,6 +157,35 @@ void testSetCustomRequestConfigNoField()
         assertEquals("nonExistentField", exception.getMessage());
     }
 
+    @Test
+    void testWaitForResponseCode()
+    {
+        SubSteps stepsToExecute = mock(SubSteps.class);
+        when(httpTestContext.getResponse()).thenReturn(new HttpResponse());
+        httpRequestSteps.waitForResponseCode(RESPONSE_CODE, Duration.ofSeconds(10), RETRY_TIMES, stepsToExecute);
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:165]: decrease duration and number of retries to some minimal value (10 seconds - too much for unit test)

File: vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java [hunk 14]
```diff
@@ -149,6 +157,35 @@ void testSetCustomRequestConfigNoField()
...
+        verify(stepsToExecute, times(RETRY_TIMES)).execute(Optional.empty());
+    }
+
+    @Test
+    void testWaitForResponseCodeWhenResponseCodeIsEqualToExpected()
+    {
+        SubSteps stepsToExecute = mock(SubSteps.class);
+        HttpResponse httpResponse = new HttpResponse();
+        httpResponse.setStatusCode(200);
+        when(httpTestContext.getResponse()).thenReturn(httpResponse);
```
Comment (valfirst) [inline vividus-plugin-rest-api/src/test/java/org/vividus/bdd/steps/api/HttpRequestStepsTests.java:184]: what about emulation of real behavior:
- 1 attempt - return http response with any code, but not 200
- 2 attempt - return http response with code 200 
?

File: vividus-tests/src/main/resources/story/integration/HttpRequest.story [hunk 15]
```diff
@@ -25,3 +25,9 @@ Then a JSON element by the JSON path '$.cookies' is equal to '{"vividus-cookie":
 Scenario: Verify HTTP cookies are cleared
 When I send HTTP GET to the relative URL '/cookies'
 Then a JSON element by the JSON path '$.cookies' is equal to '{}'
+
+Scenario: I wait for response code $responseCode for $duration duration retrying $retryTimes times $stepsToExecute
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/HttpRequest.story:29]: this test fails

File: vividus-tests/src/main/resources/story/integration/HttpRequest.story [hunk 16]
```diff
@@ -25,3 +25,9 @@ Then a JSON element by the JSON path '$.cookies' is equal to '{"vividus-cookie":
...
+When I wait for response code `200` for `PT10S` duration retrying 10 times
+|step                                                                                                                                                     |
+|When I issue a HTTP GET request for a resource with the URL '${vividus-test-site-url}/api/delayed-response?clientId=de2f4058-1a72-40dc-a6c7-ad37c6c07249'|
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/HttpRequest.story:32]: UUID should be generated at runtime, if hardcoded one is used and tests are triggered from several branches, there is a risk of collision
Vividus expression will help to generate the value: `#{generate(Internet.uuid)}`

**Pull request merged**